pkg/icingadb/sync.go

```diff
@@ -160,9 +180,61 @@ func (s Sync) ApplyDelta(ctx context.Context, delta *Delta) error {
 	// Delete
 	if len(delta.Delete) > 0 {
 		s.logger.Infof("Deleting %d items of type %s", len(delta.Delete), utils.Key(utils.Name(delta.Subject.Entity()), ' '))
-		g.Go(func() error {
-			return s.db.Delete(ctx, delta.Subject.Entity(), delta.Delete.IDs(), OnSuccessIncrement[any](stat))
-		})
+		entity := delta.Subject.Entity()
+		switch entity.(type) {
+		case *v1.Host, *v1.Service:
+			updatedCheckables := make(chan contracts.Entity, len(delta.Delete))
+			g.Go(func() error {
+				defer close(updatedCheckables)
+				s.logger.Infof("Updating %d %s sla lifecycles", len(delta.Delete), delta.Subject.Name())
+
+				first, forward, err := com.CopyFirst(ctx, CreateSlaLifecyclesFromCheckables(ctx, g, delta.Delete.Entities(ctx), true))
+				if first == nil {
+					return errors.Wrap(err, "can't copy first entity")
+				}
+				sem := s.db.GetSemaphoreForTable(utils.TableName(first))
+				stmt, _ := s.db.BuildUpdateStmt(first, []string{"delete_time"}, false)
+				if _, ok := entity.(*v1.Host); ok {
+					stmt += ` WHERE "id" = :id AND "service_id" IS NULL AND "delete_time" = 0`
+				} else {
+					stmt += ` WHERE "host_id" = :host_id AND "service_id" = :service_id AND "delete_time" = 0`
+				}
+
+				return s.db.NamedBulkExecTx(ctx, stmt, s.db.Options.MaxRowsPerTransaction, sem, forward, OnSuccessSendTo[contracts.Entity](updatedCheckables))
+			})
+
+			s.logger.Infof("Inserting %d %s sla lifecycles of type delete", len(delta.Delete), delta.Subject.Name())
+
+			entities := make(chan contracts.Entity, len(delta.Delete))
+			g.Go(func() error {
+				defer close(entities)
+
+				return s.db.CreateIgnoreStreamed(ctx, updatedCheckables, OnSuccessApplyAndSendTo(entities, GetCheckableFromSlaLifecycle))
+			})
+
+			g.Go(func() error {
+				deleteIds := make(chan interface{}, len(delta.Delete))
+
+				for {
+					entity, ok := <-entities
+					if !ok {
+						break
+					}
+
+					select {
+					case deleteIds <- entity.ID():
+					case <-ctx.Done():
+						return ctx.Err()
+					}
+				}
+
+				return s.db.DeleteStreamed(ctx, delta.Subject.Entity(), deleteIds, OnSuccessIncrement[any](stat))
+			})
```

Comment:
That looks like it should be two goroutines. Apart from that, can't what the for loop does also be done by `OnSuccessApplyAndSendTo`?

Reply:
Overseen the 🐘